From a22f4bddbcdb61ba671169d2aa0a8eea43991b61 Mon Sep 17 00:00:00 2001
From: Lyon Wang <lyon.wang@nxp.com>
Date: Fri, 30 Dec 2016 15:53:21 +0800
Subject: [PATCH 05/68] Specific patches for gstplay API

play: Add get_rotate, set_rotate API

- Add gstplay get_rotate() and set_rotate() API

play: Add force-aspect-ratio config

- Add get/set force-aspect-ratio config API

play: Add set audio / text sink API

- Add get/set audio / text sink API

Upstream-Status: Inappropriate [i.MX specific]

Signed-off-by: Lyon Wang <lyon.wang@nxp.com>

(cherry picked from commit c513c0b28326d968dd416b81b47f9e2acaa4e9a3)
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 gst-libs/gst/play/gstplay.c | 284 ++++++++++++++++++++++++++++++++++++
 gst-libs/gst/play/gstplay.h |  12 ++
 2 files changed, 296 insertions(+)

diff --git a/gst-libs/gst/play/gstplay.c b/gst-libs/gst/play/gstplay.c
index 50e9f0a..357595c 100644
--- a/gst-libs/gst/play/gstplay.c
+++ b/gst-libs/gst/play/gstplay.c
@@ -127,6 +127,7 @@ typedef enum
   CONFIG_QUARK_POSITION_INTERVAL_UPDATE,
   CONFIG_QUARK_ACCURATE_SEEK,
   CONFIG_QUARK_PIPELINE_DUMP_IN_ERROR_DETAILS,
+  CONFIG_QUARK_FORCE_ASPECT_RATIO,
 
   CONFIG_QUARK_MAX
 } ConfigQuarkId;
@@ -136,6 +137,7 @@ static const gchar *_config_quark_strings[] = {
   "position-interval-update",
   "accurate-seek",
   "pipeline-dump-in-error-details",
+  "force-aspect-ratio",
 };
 
 static GQuark _config_quark_table[CONFIG_QUARK_MAX];
@@ -321,6 +323,7 @@ gst_play_init (GstPlay * self)
       CONFIG_QUARK (POSITION_INTERVAL_UPDATE), G_TYPE_UINT, DEFAULT_POSITION_UPDATE_INTERVAL_MS,
       CONFIG_QUARK (ACCURATE_SEEK), G_TYPE_BOOLEAN, FALSE,
       CONFIG_QUARK (PIPELINE_DUMP_IN_ERROR_DETAILS), G_TYPE_BOOLEAN, FALSE,
+      CONFIG_QUARK (FORCE_ASPECT_RATIO), G_TYPE_BOOLEAN, TRUE,
       NULL);
   /* *INDENT-ON* */
 
@@ -4701,6 +4704,287 @@ gst_play_get_video_snapshot (GstPlay * self,
   return sample;
 }
 
+/**
+ * gst_get_video_sink:
+ * @play: #GstPlay instance
+ *
+ * Returns: actual video sink element
+ */
+static GstElement *
+gst_play_get_video_sink (GstPlay * self)
+{
+  GstElement *sink = NULL;
+  GstElement *actual_sink = NULL;
+  GstIteratorResult rc;
+  GstIterator *it;
+  GValue item = { 0, };
+  g_return_val_if_fail (GST_IS_PLAY (self), NULL);
+
+  g_object_get (G_OBJECT (self->playbin), "video-sink", &sink, NULL);
+  if (NULL == sink) {
+    GST_WARNING_OBJECT (self, "No video-sink found");
+    return NULL;
+  }
+  it = gst_bin_iterate_sinks ((GstBin *) sink);
+  do {
+    rc = gst_iterator_next (it, &item);
+    if (rc == GST_ITERATOR_OK) {
+      break;
+    }
+  } while (rc != GST_ITERATOR_DONE);
+
+  g_object_unref (sink);
+  actual_sink = g_value_get_object (&item);
+  g_value_unset (&item);
+  gst_iterator_free (it);
+
+  if (NULL == actual_sink) {
+    GST_WARNING_OBJECT (self, "No video-sink found");
+    return NULL;
+  }
+
+  return actual_sink;
+}
+
+/**
+ * gst_palyer_set_rotate:
+ * @play: #GstPlay instance
+ * @rotation: rotation degree value
+ *
+ * Returns: %TRUE or %FALSE
+ *
+ * Set the rotation vaule
+ */
+gboolean
+gst_play_set_rotate (GstPlay * self, gint rotation)
+{
+  GstElement *video_sink = NULL;
+  GObjectClass *gobjclass = NULL;
+  g_return_val_if_fail (GST_IS_PLAY (self), FALSE);
+
+  video_sink = gst_play_get_video_sink (self);
+  if (NULL == video_sink) {
+    GST_WARNING_OBJECT (self, " cannot get  video sink ");
+    return FALSE;
+  }
+  GST_DEBUG_OBJECT (self, "set rotation degree '%d'", rotation);
+
+  gobjclass = G_OBJECT_GET_CLASS (G_OBJECT (video_sink));
+  if (g_object_class_find_property (gobjclass, "rotate")
+      && g_object_class_find_property (gobjclass, "reconfig")) {
+    g_object_set (G_OBJECT (video_sink), "rotate", rotation / 90, NULL);
+    g_object_set (G_OBJECT (video_sink), "reconfig", 1, NULL);
+  } else if (g_object_class_find_property (gobjclass, "rotate-method")) {
+    g_object_set (G_OBJECT (video_sink), "rotate-method", rotation / 90, NULL);
+  } else {
+    GST_INFO_OBJECT (self, "can't set rotation for current video sink %s'",
+        gst_element_get_name (video_sink));
+    return FALSE;
+  }
+
+  return TRUE;
+}
+
+/**
+ * gst_play_get_rotate:
+ * @play: #GstPlay instance
+ *
+ * Returns: the rotation degree value
+ */
+gint
+gst_play_get_rotate (GstPlay * self)
+{
+  GstElement *video_sink = NULL;
+  GObjectClass *gobjclass = NULL;
+  gint rotation = 0;
+  g_return_val_if_fail (GST_IS_PLAY (self), 0);
+
+  video_sink = gst_play_get_video_sink (self);
+  if (NULL == video_sink) {
+    GST_WARNING_OBJECT (self, " cannot get  video sink ");
+    return 0;
+  }
+
+  /* check if the element has "rotate" property */
+  gobjclass = G_OBJECT_GET_CLASS (video_sink);
+  if (g_object_class_find_property (gobjclass, "rotate")) {
+    g_object_get (G_OBJECT (video_sink), "rotate", &rotation, NULL);
+    rotation = rotation * 90;
+  } else if (g_object_class_find_property (gobjclass, "rotate-method")) {
+    g_object_get (G_OBJECT (video_sink), "rotate-method", &rotation, NULL);
+    rotation = rotation * 90;
+  }
+
+  GST_DEBUG_OBJECT (self, "get rotation degree '%d'", rotation);
+
+  return rotation;
+}
+
+/**
+ * gst_play_config_set_force_aspect_ratio:
+ * @play: #GstPlay instance
+ * @force_aspect_ratio: keey original aspect ratio or not
+ *
+ * Enable or disable force aspect ratio
+ * force_aspect_ratio seeking is TRUE by default.
+ *
+ * Since: 1.12
+ */
+void
+gst_play_config_set_force_aspect_ratio (GstPlay * self, gboolean force_aspect_ratio)
+{
+  GstStructure *config = self->config;
+  g_return_if_fail (config != NULL);
+
+  gst_structure_id_set (config,
+      CONFIG_QUARK (FORCE_ASPECT_RATIO), G_TYPE_BOOLEAN, force_aspect_ratio, NULL);
+
+  g_object_set(self->playbin, "force-aspect-ratio", force_aspect_ratio, NULL);
+}
+
+/**
+ * gst_play_config_get_force_aspect_ratio:
+ * @config: a #GstPlay configuration
+ *
+ * Returns: %TRUE if force-aspect-ratio is enabled
+ *
+ * Since 1.12
+ */
+gboolean
+gst_play_config_get_force_aspect_ratio (const GstStructure * config)
+{
+  gboolean force_aspect_ratio = TRUE;
+
+  g_return_val_if_fail (config != NULL, FALSE);
+
+  gst_structure_id_get (config,
+      CONFIG_QUARK (FORCE_ASPECT_RATIO), G_TYPE_BOOLEAN, &force_aspect_ratio, NULL);
+
+  return force_aspect_ratio;
+}
+
+/**
+ * gst_play_set_audio_sink:
+ * @play: #GstPlay instance
+ * @audio_sink: the custom audio sink to set
+ *
+ * Returns: %TRUE or %FALSE
+ *
+ * Set the customize audio sink
+ */
+gboolean
+gst_play_set_audio_sink (GstPlay * self,  GstElement * audio_sink)
+{
+  g_return_val_if_fail (GST_IS_PLAY (self), FALSE);
+  g_return_val_if_fail (audio_sink != NULL, FALSE);
+
+  g_object_set (G_OBJECT (self->playbin), "audio-sink", audio_sink, NULL);
+  return TRUE;
+}
+
+/**
+ * gst_play_set_text_sink:
+ * @play: #GstPlay instance
+ * @text_sink: the custom text sink  to set
+ *
+ * Returns: %TRUE or %FALSE
+ *
+ * Set the customize text sink
+ */
+gboolean
+gst_play_set_text_sink (GstPlay * self,  GstElement * text_sink)
+{
+  g_return_val_if_fail (GST_IS_PLAY (self), FALSE);
+  g_return_val_if_fail (text_sink != NULL, FALSE);
+
+  g_object_set (G_OBJECT (self->playbin), "text-sink", text_sink, NULL);
+  return TRUE;
+}
+
+/**
+ * gst_play_get_audio_sink:
+ * @play: #GstPlay instance
+ *
+ * Returns: actual audio sink element
+ */
+GstElement *
+gst_play_get_audio_sink (GstPlay * self)
+{
+  GstElement *sink = NULL;
+  GstElement *actual_sink = NULL;
+  GstIteratorResult rc;
+  GstIterator *it;
+  GValue item = { 0, };
+  g_return_val_if_fail (GST_IS_PLAY (self), NULL);
+
+  g_object_get (G_OBJECT (self->playbin), "audio-sink", &sink, NULL);
+  if (NULL == sink) {
+    GST_WARNING_OBJECT (self, "No audio-sink found");
+    return NULL;
+  }
+  it = gst_bin_iterate_sinks ((GstBin *) sink);
+  do {
+    rc = gst_iterator_next (it, &item);
+    if (rc == GST_ITERATOR_OK) {
+      break;
+    }
+  } while (rc != GST_ITERATOR_DONE);
+
+  g_object_unref (sink);
+  actual_sink = g_value_get_object (&item);
+  g_value_unset (&item);
+  gst_iterator_free (it);
+
+  if (NULL == actual_sink) {
+    GST_WARNING_OBJECT (self, "No auido-sink found");
+    return NULL;
+  }
+
+  return actual_sink;
+}
+
+/**
+ * gst_play_get_text_sink:
+ * @play: #GstPlay instance
+ *
+ * Returns: actual text sink element
+ */
+GstElement *
+gst_play_get_text_sink (GstPlay * self)
+{
+  GstElement *sink = NULL;
+  GstElement *actual_sink = NULL;
+  GstIteratorResult rc;
+  GstIterator *it;
+  GValue item = { 0, };
+  g_return_val_if_fail (GST_IS_PLAY (self), NULL);
+
+  g_object_get (G_OBJECT (self->playbin), "text-sink", &sink, NULL);
+  if (NULL == sink) {
+    GST_WARNING_OBJECT (self, "No text-sink found");
+    return NULL;
+  }
+  it = gst_bin_iterate_sinks ((GstBin *) sink);
+  do {
+    rc = gst_iterator_next (it, &item);
+    if (rc == GST_ITERATOR_OK) {
+      break;
+    }
+  } while (rc != GST_ITERATOR_DONE);
+
+  g_object_unref (sink);
+  actual_sink = g_value_get_object (&item);
+  g_value_unset (&item);
+  gst_iterator_free (it);
+
+  if (NULL == actual_sink) {
+    GST_WARNING_OBJECT (self, "No text-sink found");
+    return NULL;
+  }
+
+  return actual_sink;
+}
+
 /**
  * gst_play_is_play_message:
  * @msg: A #GstMessage
diff --git a/gst-libs/gst/play/gstplay.h b/gst-libs/gst/play/gstplay.h
index 825f6e7..7ee5c69 100644
--- a/gst-libs/gst/play/gstplay.h
+++ b/gst-libs/gst/play/gstplay.h
@@ -444,6 +444,18 @@ void           gst_play_message_parse_volume_changed             (GstMessage *ms
 GST_PLAY_API
 void           gst_play_message_parse_muted_changed              (GstMessage *msg, gboolean *muted);
 
+/* Custom gstplay API */
+gboolean    gst_play_set_rotate (GstPlay * play, gint rotation);
+gint        gst_play_get_rotate (GstPlay * play);
+
+void        gst_play_config_set_force_aspect_ratio (GstPlay * self, gboolean force_aspect_ratio);
+gboolean    gst_play_config_get_force_aspect_ratio (const GstStructure * config);
+
+gboolean    gst_play_set_audio_sink (GstPlay * play, GstElement * audio_sink);
+gboolean    gst_play_set_text_sink (GstPlay * play, GstElement * text_sink);
+GstElement * gst_play_get_audio_sink (GstPlay * play);
+GstElement * gst_play_get_text_sink (GstPlay * play);
+
 G_END_DECLS
 
 #endif /* __GST_PLAY_H__ */
-- 
2.43.0

